File: python/features/steps/examples_steps.py
"""Steps for example project integration setup."""

from __future__ import annotations

import shutil
from pathlib import Path

from behave import given, then, when

from features.steps.shared import ensure_git_repository, run_cli
from taskulus.agents_management import (
    build_project_management_text,
    TASKULUS_SECTION_TEXT,
)


README_STUB = "This is a sample project that uses Taskulus."


def _examples_root() -> Path:
    return Path(__file__).resolve().parents[3] / "examples"


def _example_dir(name: str) -> Path:
    slug = name.strip().lower().replace(" ", "-")
    return _examples_root() / slug


@given('the "{name}" example project does not exist')
def given_example_missing(context: object, name: str) -> None:
    path = _example_dir(name)
    if path.exists():
        shutil.rmtree(path)


@when('I create the "{name}" example project')
def when_create_example_project(context: object, name: str) -> None:
    path = _example_dir(name)
    path.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(path)


@when('I run "tsk setup agents" in the "{name}" example project')
def when_run_setup_agents_in_example(context: object, name: str) -> None:
    path = _example_dir(name)
    context.working_directory = path
    run_cli(context, "tsk setup agents")


@when('I run "tsk init" in the "{name}" example project')
def when_run_init_in_example(context: object, name: str) -> None:
    path = _example_dir(name)
    context.working_directory = path
    run_cli(context, "tsk init")


@when('I add a README stub to the "{name}" example project')
def when_add_readme_stub(context: object, name: str) -> None:
    path = _example_dir(name)
    readme = path / "README.md"
    readme.write_text(README_STUB + "\n", encoding="utf-8")


@then('the "{name}" example project should contain a README stub')
def then_example_contains_readme(context: object, name: str) -> None:
    path = _example_dir(name)
    readme = path / "README.md"
    assert readme.exists()
    content = readme.read_text(encoding="utf-8").strip()
    assert content == README_STUB


@then('the "{name}" example project should contain .taskulus.yml')
def then_example_contains_config(context: object, name: str) -> None:
    path = _example_dir(name)
    assert (path / ".taskulus.yml").exists()


@then('the "{name}" example project should contain the project directory')
def then_example_contains_project_dir(context: object, name: str) -> None:
    path = _example_dir(name)
    assert (path / "project").exists()


@then(
    'the "{name}" example project should contain AGENTS.md with Taskulus instructions'
)
def then_example_contains_agents(context: object, name: str) -> None:
    path = _example_dir(name)
    agents = path / "AGENTS.md"
    assert agents.exists()
    content = agents.read_text(encoding="utf-8")
    assert TASKULUS_SECTION_TEXT.strip() in content


@then('the "{name}" example project should contain CONTRIBUTING_AGENT.md')
def then_example_contains_instructions(context: object, name: str) -> None:
    path = _example_dir(name)
    instructions = path / "CONTRIBUTING_AGENT.md"
    assert instructions.exists()
    content = instructions.read_text(encoding="utf-8")
    expected = build_project_management_text(path)
    assert content.strip() == expected.strip()

File: rust/features/steps/examples_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use cucumber::{given, then, when};

use kanbus::agents_management::{kanbus_section_text, project_management_text};

use crate::step_definitions::initialization_steps::KanbusWorld;

const README_STUB: &str = "This is a sample project that uses Kanbus.";

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..")
}

fn example_dir(name: &str) -> PathBuf {
    let slug = name.trim().to_lowercase().replace(' ', "-");
    repo_root().join("examples").join(slug)
}

#[given(expr = "the {string} example project does not exist")]
fn given_example_missing(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    if path.exists() {
        fs::remove_dir_all(&path).expect("remove example dir");
    }
}

#[when(expr = "I create the {string} example project")]
fn when_create_example_project(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    fs::create_dir_all(&path).expect("create example dir");
    let status = Command::new("git")
        .args(["init"])
        .current_dir(&path)
        .status()
        .expect("git init failed");
    if !status.success() {
        panic!("git init failed");
    }
}

#[when(expr = "I add a README stub to the {string} example project")]
fn when_add_readme_stub(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    let readme = path.join("README.md");
    fs::write(readme, format!("{README_STUB}\n")).expect("write README");
}

#[then(expr = "the {string} example project should contain a README stub")]
fn then_example_contains_readme(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    let readme = path.join("README.md");
    assert!(readme.exists());
    let content = fs::read_to_string(readme).expect("read README");
    assert_eq!(content.trim(), README_STUB);
}

#[then(expr = "the {string} example project should contain .kanbus.yml")]
fn then_example_contains_config(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    assert!(path.join(".kanbus.yml").exists());
}

#[then(expr = "the {string} example project should contain the project directory")]
fn then_example_contains_project_dir(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    assert!(path.join("project").exists());
}

#[then(expr = "the {string} example project should contain AGENTS.md with Kanbus instructions")]
fn then_example_contains_agents(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    let agents = path.join("AGENTS.md");
    assert!(agents.exists());
    let content = fs::read_to_string(agents).expect("read AGENTS.md");
    assert!(content.contains(kanbus_section_text().trim()));
}

#[then(expr = "the {string} example project should contain CONTRIBUTING_AGENT.md")]
fn then_example_contains_instructions(_world: &mut KanbusWorld, name: String) {
    let path = example_dir(&name);
    let instructions = path.join("CONTRIBUTING_AGENT.md");
    assert!(instructions.exists());
    let content = fs::read_to_string(instructions).expect("read instructions");
    let expected = project_management_text(&path).expect("instructions text");
    assert_eq!(content.trim(), expected.trim());
}
